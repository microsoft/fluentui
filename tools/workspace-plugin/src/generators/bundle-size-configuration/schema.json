{
  "$schema": "http://json-schema.org/schema",
  "$id": "BundleSizeConfiguration",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Project for which to generate bundle-size configuration.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "For which project do you want to generate bundle-size configuration?",
      "x-dropdown": "projects"
    },
    "overrideBaseConfig": {
      "type": "boolean",
      "description": "generate monosize config within project in order to add custom bundling overrides",
      "x-prompt": "Would you like to override monosize bundling config for measurements?",
      "default": false
    }
  },
  "required": ["name"]
}
