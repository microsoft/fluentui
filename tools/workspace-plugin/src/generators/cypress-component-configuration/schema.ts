export interface CypressComponentConfigurationGeneratorSchema {
  /**
   * The name of the project to add cypress component testing to
   */
  project: string;
}
