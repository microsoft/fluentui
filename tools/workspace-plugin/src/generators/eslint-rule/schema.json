{
  "$schema": "http://json-schema.org/schema",
  "$id": "EslintRule",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the new rule.",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "directory": {
      "type": "string",
      "description": "Create the rule under this directory within `tools/eslint-rules/` (can be nested).",
      "alias": "dir",
      "default": "rules"
    }
  },
  "required": ["name", "directory"]
}
