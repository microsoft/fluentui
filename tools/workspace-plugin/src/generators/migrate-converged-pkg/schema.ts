export interface MigrateConvergedPkgGeneratorSchema {
  /**
   * Library name or comma delimited library names to execute migration on multiple libraries.
   */
  name?: string;
  /**
   * Get statistics for how many projects have been migrated
   */
  stats?: boolean;
  /**
   * Run generator on all vNext packages
   */
  all?: boolean;
  /**
   * Add particular team to CODEOWNERS file
   */
  owner?: string;
}
