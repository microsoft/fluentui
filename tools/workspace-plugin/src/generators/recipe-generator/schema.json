{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "id": "recipe-generator",
  "type": "object",
  "properties": {
    "recipeName": {
      "type": "string",
      "description": "Recipe name",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Please enter the recipe name"
    }
  },
  "required": ["recipeName"]
}
