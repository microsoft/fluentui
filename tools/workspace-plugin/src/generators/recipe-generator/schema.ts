export interface RecipeGeneratorGeneratorSchema {
  /**
   * Recipe name
   */
  recipeName: string;
}
