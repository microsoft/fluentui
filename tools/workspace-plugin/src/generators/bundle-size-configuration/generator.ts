import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  visitNotIgnoredFiles,
} from '@nx/devkit';

import * as path from 'path';

import { PackageJson } from '../../types';
import { assertStoriesProject, isSplitProject } from '../split-library-in-two/shared';

import { BundleSizeConfigurationGeneratorSchema } from './schema';

export async function bundleSizeConfigurationGenerator(tree: Tree, schema: BundleSizeConfigurationGeneratorSchema) {
  const options = normalizeOptions(tree, schema);

  const project = readProjectConfiguration(tree, options.name);

  assertOptions(tree, { isSplitProject: isSplitProject(tree, project), project });

  const configPaths = {
    bundleSizeRoot: joinPathFragments(project.root, 'bundle-size'),
    bundleSizeConfig: joinPathFragments(project.root, 'monosize.config.mjs'),
  };

  generateFiles(tree, path.join(__dirname, 'files'), project.root, {
    packageName: options.name,
    rootOffset: offsetFromRoot(project.root),
  });

  let hasFixtures = false;
  visitNotIgnoredFiles(tree, configPaths.bundleSizeRoot, filePath => {
    if (!filePath.endsWith('index.fixture.js') && filePath.endsWith('.fixture.js')) {
      hasFixtures = true;
    }
  });

  if (hasFixtures) {
    tree.delete(joinPathFragments(configPaths.bundleSizeRoot, 'index.fixture.js'));
  }

  if (options.overrideBaseConfig === false) {
    tree.delete(configPaths.bundleSizeConfig);
  }

  updateJson(tree, joinPathFragments(project.root, 'package.json'), (json: PackageJson) => {
    json.scripts = json.scripts ?? {};
    json.scripts['bundle-size'] = 'monosize measure';
    return json;
  });

  await formatFiles(tree);
}

function normalizeOptions(tree: Tree, schema: BundleSizeConfigurationGeneratorSchema) {
  return {
    overrideBaseConfig: false,
    ...schema,
  };
}

function assertOptions(tree: Tree, options: { project: ProjectConfiguration; isSplitProject: boolean }) {
  assertStoriesProject(tree, options);
}

export default bundleSizeConfigurationGenerator;
