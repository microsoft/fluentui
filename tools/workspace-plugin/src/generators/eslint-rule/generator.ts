import { formatFiles, joinPathFragments, ProjectConfiguration, readProjectConfiguration, Tree } from '@nx/devkit';

import * as tsquery from '@phenomnomnominal/tsquery';
import { EslintRuleGeneratorSchema } from './schema';
import { lintWorkspaceRuleGenerator } from '@nx/eslint/src/generators/workspace-rule/workspace-rule';

interface Options extends ReturnType<typeof normalizeSchema> {}

export async function eslintRuleGenerator(tree: Tree, schema: EslintRuleGeneratorSchema) {
  const options = normalizeSchema(schema);
  await lintWorkspaceRuleGenerator(tree, options);

  const project = readProjectConfiguration(tree, 'eslint-rules');
  replaceBoilerplateToUseOldTypescriptEslintSyntax(tree, { project, ...options });

  await formatFiles(tree);
}

export default eslintRuleGenerator;

function normalizeSchema(schema: EslintRuleGeneratorSchema) {
  return { ...schema, directory: schema.directory ?? 'rules' } as Required<EslintRuleGeneratorSchema>;
}

// TODO: remove this after we migrate to @typescript-eslint v5 which exposes utils package instead of experimental-utils
function replaceBoilerplateToUseOldTypescriptEslintSyntax(
  tree: Tree,
  options: Options & { project: ProjectConfiguration },
) {
  const ruleDir = joinPathFragments(options.project.root, options.directory);
  const paths = {
    impl: joinPathFragments(ruleDir, `${options.name}.ts`),
    spec: joinPathFragments(ruleDir, `${options.name}.spec.ts`),
  };

  const implContent = tree.read(paths.impl, 'utf-8') ?? '';
  const specContent = tree.read(paths.spec, 'utf-8') ?? '';

  const defaultImport = '@typescript-eslint/utils';
  const ourImport = '@typescript-eslint/experimental-utils';
  const IMPORT_AST_SELECTOR = `ImportDeclaration StringLiteral[value=${defaultImport}]`;

  const updatedSpecContent = tsquery.replace(specContent, IMPORT_AST_SELECTOR, () => {
    return `'${ourImport}'`;
  });

  let updatedImplContent = tsquery.replace(implContent, IMPORT_AST_SELECTOR, () => {
    return `'${ourImport}'`;
  });
  updatedImplContent = tsquery.replace(
    updatedImplContent,
    'ObjectLiteralExpression PropertyAssignment Identifier[name=docs] ~ ObjectLiteralExpression',
    () => {
      return `{
        category: 'Best Practices',
        description: \`\`,
        recommended: 'error',
      }`;
    },
  );

  tree.write(paths.impl, updatedImplContent);
  tree.write(paths.spec, updatedSpecContent);
}
