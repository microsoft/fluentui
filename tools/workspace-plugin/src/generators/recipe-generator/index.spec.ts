import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { addProjectConfiguration, Tree, writeJson } from '@nx/devkit';
import path from 'path';

import generator from './index';
import { RecipeGeneratorGeneratorSchema } from './schema';

const recipesRoot = 'apps/recipes-react-components/src/recipes';
const recipeName = 'Hello World';
const recipePackageName = 'hello-world';

const setup = (tree: Tree) => {
  const generateApp = () => {
    const paths = { root: 'apps/recipes-react-components' };
    writeJson(tree, path.join(paths.root, 'package.json'), {
      name: '@proj/recipes-react-components',
      private: true,
    });
    tree.write(path.join(paths.root, 'src/recipes/.gitkeep'), '');

    addProjectConfiguration(tree, '@proj/recipes-react-components', {
      root: paths.root,
      sourceRoot: path.join(paths.root, 'src'),
      projectType: 'application',
      targets: {},
    });
  };

  generateApp();
  return tree;
};

describe('recipe-generator generator', () => {
  let tree: Tree;
  const options: RecipeGeneratorGeneratorSchema = { recipeName };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    tree = setup(tree);
  });

  it('should generate boilerplate', async () => {
    await generator(tree, options);

    ['HelloWorld.stories.mdx', 'code-snippets/HelloWorld.tsx', 'code-snippets/index.ts', 'code-snippets', ''].forEach(
      file => {
        expect(tree.exists(path.join(recipesRoot, recipePackageName, file))).toBeTruthy();
      },
    );
  });

  it('should generate implementation boilerplate', async () => {
    await generator(tree, options);

    const storyContent = tree.read(path.join(recipesRoot, recipePackageName, 'HelloWorld.stories.mdx'), 'utf-8');
    const codeSnippetContent = tree.read(
      path.join(recipesRoot, recipePackageName, 'code-snippets/HelloWorld.tsx'),
      'utf-8',
    );
    const codeSnippetIndexContent = tree.read(
      path.join(recipesRoot, recipePackageName, 'code-snippets/index.ts'),
      'utf-8',
    );

    expect(codeSnippetContent).toMatchInlineSnapshot(`
      "import * as React from 'react';
      import { Button } from '@fluentui/react-components';

      export const HelloWorldSnippetExample = () => {
        return <Button>This is a Button</Button>;
      };
      "
    `);
    expect(codeSnippetIndexContent).toMatchInlineSnapshot(`
      "export * from './HelloWorld';
      "
    `);
    expect(storyContent).toMatchSnapshot();
  });
});
