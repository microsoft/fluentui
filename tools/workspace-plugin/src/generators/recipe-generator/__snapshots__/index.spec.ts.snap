// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipe-generator generator should generate implementation boilerplate 3`] = `
"import LinkTo from '@storybook/addon-links/react';
import { TemplateExample } from '../../templates';
import { Meta } from '@storybook/addon-docs';
import { HelloWorldSnippetExample } from './code-snippets';

<Meta title=\\"Concepts/Recipes/Hello World\\" />

---

# Hello World recipe

### **Overview**

<!--
Add a description of the recipe here. This should be a short description and what it does.
-->

### **Ingredients**

<!--
Add a list of ingredients here. This should be a list of required/needed dependencies such as: @fluentui/react-button.
-->

## **Steps**

<!--
Add a list of steps here. This should be an explanation of how to achieve the outcome.
Note: Code snippets in this section should be minimal and only include the important parts of the component.
-->

This is an example of a code snippet:

<TemplateExample>
  <HelloWorldSnippetExample />
</TemplateExample>

\`\`\`tsx
<Button>This is a Button</Button>
\`\`\`

## **Variants**

<!--
Add a list of variants here, not all recipes will have variants.
When adding examples, make sure to add a description of the variant and preferably explain the main idea over providing code snippets.
-->

## **Best practices**

<!--
Add a list of best practices here. This list could have do's and don'ts or just general guidelines.
-->
"
`;
